### PR TITLE
Throw ConfigException when process fails with "com.jcraft.jsch.JSchException: Auth fail"

### PR DESCRIPTION
**Pull request by sakama**
I noticed this plugin throws `org.apache.commons.vfs2.FileSystemException` when plugin fails due to wrong credentials.
In this case, plugin should throw `org.embulk.config.ConfigException`.
This makes difference when user calls Embulk from EmbulkEmbed.
Requester never knows this failure is retryable or not.

I changed implementation to throw ConfigException.
Generally, ConfigException should throw when first attempt fails. But SFTP library seems bit unstable. So that I make plugin to throw ConfigException when last retry is failed.

```bash
2018-04-18 10:14:56.591 +0000 [WARN] (0001:transaction): SFTP GET request failed. Retrying 2/5 after 1 seconds. Message: Could not connect to SFTP server at "sftp://sftp_user:***@example.com/".
2018-04-18 10:14:57.591 +0000 [INFO] (0001:transaction): Getting to download file list
2018-04-18 10:14:57.599 +0000 [INFO] (0001:transaction): Connecting to sftp://sftp_user:***@example.com:22/
2018-04-18 10:15:23.596 +0000 [WARN] (0001:transaction): SFTP GET request failed. Retrying 3/5 after 2 seconds. Message: Could not connect to SFTP server at "sftp://sftp_user:***@example.com/".
org.apache.commons.vfs2.FileSystemException: Could not connect to SFTP server at "sftp://sftp_user:***@example.com/".
	at org.apache.commons.vfs2.provider.sftp.SftpFileProvider.doCreateFileSystem(SftpFileProvider.java:86) ~[na:na]
	at org.apache.commons.vfs2.provider.AbstractOriginatingFileProvider.getFileSystem(AbstractOriginatingFileProvider.java:92) ~[na:na]
	at org.apache.commons.vfs2.provider.AbstractOriginatingFileProvider.findFile(AbstractOriginatingFileProvider.java:71) ~[na:na]
	at org.apache.commons.vfs2.provider.AbstractOriginatingFileProvider.findFile(AbstractOriginatingFileProvider.java:55) ~[na:na]
	at org.apache.commons.vfs2.impl.DefaultFileSystemManager.resolveFile(DefaultFileSystemManager.java:711) ~[na:na]
	at org.apache.commons.vfs2.impl.DefaultFileSystemManager.resolveFile(DefaultFileSystemManager.java:648) ~[na:na]
	at org.embulk.input.sftp.SftpFileInput$1.call(SftpFileInput.java:207) ~[na:na]
	at org.embulk.input.sftp.SftpFileInput$1.call(SftpFileInput.java:187) ~[na:na]
	at org.embulk.spi.util.RetryExecutor.run(RetryExecutor.java:100) [embulk-core-0.8.36.jar:na]
	at org.embulk.spi.util.RetryExecutor.runInterruptible(RetryExecutor.java:77) [embulk-core-0.8.36.jar:na]
	at org.embulk.input.sftp.SftpFileInput.listFilesByPrefix(SftpFileInput.java:187) [embulk-input-sftp-0.2.8.jar:na]
	at org.embulk.input.sftp.SftpFileInputPlugin.transaction(SftpFileInputPlugin.java:22) [embulk-input-sftp-0.2.8.jar:na]
	at org.embulk.spi.FileInputRunner.transaction(FileInputRunner.java:67) [embulk-core-0.8.36.jar:na]
	...
Caused by: org.apache.commons.vfs2.FileSystemException: Could not connect to SFTP server at "example.com".
	at org.apache.commons.vfs2.provider.sftp.SftpClientFactory.createConnection(SftpClientFactory.java:147) ~[na:na]
	at org.apache.commons.vfs2.provider.sftp.SftpFileProvider.doCreateFileSystem(SftpFileProvider.java:79) ~[na:na]
	... 28 common frames omitted
Caused by: com.jcraft.jsch.JSchException: Auth fail
	at com.jcraft.jsch.Session.connect(Session.java:519) ~[jsch-0.1.42.jar:na]
	at com.jcraft.jsch.Session.connect(Session.java:183) ~[jsch-0.1.42.jar:na]
	at org.apache.commons.vfs2.provider.sftp.SftpClientFactory.createConnection(SftpClientFactory.java:145) ~[na:na]
	... 29 common frames omitted
```

### CODE REVIEW

**Comment by sakama** [comment]
Ah, CI failed. I'll fix it.

**Comment by sakama** [comment]
CI retry fixed CI failure.
But I also created another PR to use only Java 8. #30

**Comment by sakama** [comment]
@tvhung83 Could you review this?

File: src/main/java/org/embulk/input/sftp/SftpFileInput.java [hunk 1]
```diff
@@ -264,6 +265,14 @@ public void onRetry(Exception exception, int retryCount, int retryLimit, int ret
                         public void onGiveup(Exception firstException, Exception lastException)
                                 throws RetryGiveupException
                         {
+                            // Generally, Auth fail should be caught and throw ConfigException when first retry. But this library is a bit unstable.
```
Comment (tvhung83) [inline src/main/java/org/embulk/input/sftp/SftpFileInput.java:268]: I was not sure if we should retry this, but then saw your comments, so 👍

File: src/main/java/org/embulk/input/sftp/SftpFileInput.java [hunk 2]
```diff
@@ -264,6 +265,14 @@ public void onRetry(Exception exception, int retryCount, int retryLimit, int ret
...
+                            // So we throw ConfigException after all retries are completed
+                            if (lastException.getCause().getCause() != null) {
```
Comment (tvhung83) [inline src/main/java/org/embulk/input/sftp/SftpFileInput.java:270]: ~This looks brittle. Maybe we should trace to root cause?~ I took a quick look into `commons-vfs2` source code, it seems to be good enough.

**Review by tvhung83 — APPROVED** [review]
LGTM

**Pull request merged**